x-pack/plugins/transform/public/app/hooks/use_index_data.ts

```diff
@@ -46,9 +45,62 @@ export const useIndexData = (
     },
   } = useAppDependencies();
 
-  const indexPatternFields = getFieldsFromKibanaIndexPattern(indexPattern);
+  const [indexPatternFields, setIndexPatternFields] = useState<string[]>();
+
+  // Fetch 500 random documents to determine populated fields.
+  // This is a workaround to avoid passing potentially thousands of unpopulated fields
+  // (for example, as part of filebeat/metricbeat/ECS based indices)
+  // to the data grid component which would significantly slow down the page.
+  const fetchDataGridSampleDocuments = async function () {
+    setErrorMessage('');
+    setStatus(INDEX_STATUS.LOADING);
+
+    const esSearchRequest = {
+      index: indexPattern.title,
+      body: {
+        fields: ['*'],
+        _source: false,
+        query: {
+          function_score: {
+            query: { match_all: {} },
+            random_score: {},
+          },
+        },
+        size: 500,
+      },
+    };
+
+    const resp = await api.esSearch(esSearchRequest);
+
+    if (!isEsSearchResponse(resp)) {
+      setErrorMessage(getErrorMessage(resp));
+      setStatus(INDEX_STATUS.ERROR);
+      return;
+    }
+
+    const isCrossClusterSearch = indexPattern.title.includes(':');
+    const isMissingFields = resp.hits.hits.every((d) => typeof d.fields === 'undefined');
+
+    const docs = resp.hits.hits.map((d) => getProcessedFields(d.fields ?? {}));
+    // Get all field names for each returned doc and flatten it
+    // to a list of unique field names used across all docs.
+    const populatedFields = [...new Set(docs.map(Object.keys).flat(1))];
+
+    setCcsWarning(isCrossClusterSearch && isMissingFields);
+    setStatus(INDEX_STATUS.LOADED);
+    setIndexPatternFields(populatedFields);
+  };
+
+  useEffect(() => {
+    fetchDataGridSampleDocuments();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
```

Comment:
I'm assuming we don't need indexPattern here as a dependency?

Reply:
We only need to fetch it once on load since the index pattern cannot change, it's hard coded in the URL for transforms.